### PR TITLE
Feature/register plan explainer

### PR DESCRIPTION
**Pull request by goedderz**
### Scope & Purpose

**Register Planning Update:** This PR fixes some issues we found in the `RegisterPlan.cpp`. Additionally we optimized the register assignment even further (using information that initially were used by the explainer).

Explainer for the register plan of a query. Currently targeted for internal use only.

Can be executed in `arangosh` (or `arangod --console`) by calling the function `require("@arangodb/aql/explainer").explainRegisters`. Its used exactly like `require("@arangodb/aql/explainer").explain`. For example:

```
$ arangosh --javascript.execute-string 'require("@arangodb/aql/explainer").explainRegisters({query: "FOR i IN 1..3 LET x = (RETURN 2*i) RETURN x", options: {optimizer: {rules: ["-splice-subqueries"]}}})'
```

- [X] Strictly *new functionality* (and just internal)

Note that you have to understand some details of the register planning and AQL execution to make sense of this.

Things left **TODO** here:

- [ ] Add an (also undocumented) option to the explainer which adds this to the usual output.
- [X] Add the register information necessary for this to debugDumps (by just setting the option `explainRegisters: true`).
- [ ] Make it possible to explain the register plan given a debug dump.
- [ ] ~Add a CHANGELOG entry for the functional changes, and if feasible move them to a separate PR.~
- [ ] As this can be easily found via auto completion, add some documentation, at least that this is for internal debugging and subject to change or be removed without notice.

### CODE REVIEW

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/10194/

**Comment by joerg84** [comment]
Is there a corresponding docs PR?

**Comment by goedderz** [comment]
> Is there a corresponding docs PR?

Purposefully not. This is both more suited for and more of interest to developers than users.

File: tests/IResearch/IResearchQueryJoin-test.cpp [hunk 1]
```diff
@@ -1496,13 +1496,8 @@ TEST_F(IResearchQueryJoinTest, test) {
...
         "customscorer(d, x.seq) DESC RETURN d";
 
-    EXPECT_TRUE(arangodb::tests::assertRules(vocbase, query,
-                                             {
-                                                 arangodb::aql::OptimizerRule::handleArangoSearchViewsRule,
-                                             }));
-
     auto queryResult = arangodb::tests::executeQuery(vocbase, query);
-    ASSERT_TRUE(queryResult.result.is(TRI_ERROR_BAD_PARAMETER));
+    ASSERT_TRUE(queryResult.result.is(TRI_ERROR_INTERNAL));
```
Comment (gnusi) [inline tests/IResearch/IResearchQueryJoin-test.cpp:1500]: This is breaking change. We have to understand why `handleArangoSearchViewsRule` is not applied anymore and avoid changing error code.

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/10301/

**Comment by goedderz** [comment]
Yes, and I know what's going on. Will fix it.

**Comment by maierlars** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/10341/

**Comment by maierlars** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/10354/

**Comment by maierlars** [comment]
image

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10551/

**Comment by jsteemann** [comment]
Tests blue. Waiting for feedback from @gnusi

**Pull request merged**